CorsixTH/Lua/research_department.lua

```diff
@@ -400,30 +400,69 @@ end
 !param drug The drug to improve, table taken from world.available_diseases
 --]]
 function ResearchDepartment:improveDrug(drug)
-  local research_info = self.research_progress[drug]
   local disease = self.hospital.disease_casebook[drug.disease.id]
+  local level_config = self.world.map.level_config
+  local min_drug_cost = level_config.gbv.MinDrugCost
+
+  -- Work out the improvement to make
+  local function decideImprovement()
+    local at_max_effectiveness = disease.cure_effectiveness >= 100
+    local at_min_cost = disease.drug_cost <= min_drug_cost
+    local improvement
+    if at_max_effectiveness and at_min_cost then -- Nothing left to improve
+      return false
+    end
+    if at_max_effectiveness then -- Only cost improvement remains
+      improvement = "cost"
+    elseif at_min_cost then -- Only effectiveness improvement remains
+      improvement =  "effectiveness"
+    else -- Either can be improved
+      improvement = math.random(1, 2) == 1 and "cost" or "effectiveness"
+      -- A 1/7 chance both are improved
+      improvement = math.random(1, 7) == 1 and "both" or improvement
+    end
+    return improvement
+  end
 
-  -- Improving effectiveness and cost should alternate
-  if research_info.effect_imp > research_info.cost_imp then
-    -- Time to improve cost
-    disease.drug_cost = disease.drug_cost - 10
-    research_info.cost_imp = research_info.cost_imp + 1
-    if disease.cure_effectiveness == 100 then
+  -- Decrease the cost of the drug
+  -- Set to 10%, provides good balance vs number of steps on the effectiveness ladder
+  local function decreaseDrugCost()
+    local new_cost = math.max(min_drug_cost, math.floor(disease.drug_cost * 0.9))
+    disease.drug_cost = new_cost
+    if self.research_policy.drugs.current == drug then
+      self:nextResearch("drugs")
+    end
```

Comment:
I like these changes!

To keep this function cleanly implement "decrease costs" only, you can move these lines after the call at 462.

I agree it's weird and probably wrong, but I agree with you that it's better in a next patch.

Reply:
This one has been done